README.md
[![Module Tests](https://github.com/NicTool/dns-resource-record/actions/workflows/ci-test.yml/badge.svg)](https://github.com/NicTool/dns-resource-record/actions/workflows/ci-test.yml)
[![Coverage Status](https://coveralls.io/repos/github/NicTool/dns-resource-record/badge.svg?branch=master)](https://coveralls.io/github/NicTool/dns-resource-record?branch=master)

# dns-resource-record

DNS resource record parser, validator, importer, and exporter.


## SYNOPSIS

This module is used to:

- validate well formedness and RFC compliance of DNS resource records
- import RRs from:
    - JS object
    - JSON
    - [BIND](https://www.isc.org/bind/) zone [file format](https://bind9.readthedocs.io/en/latest/reference.html#zone-file)
    - tinydns [data format](https://cr.yp.to/djbdns/tinydns-data.html)
- export RRs to:
    - BIND zone files
    - tinydns data
    - maradns
    - JS object
    - JSON

This package intends to import and export RFC compliant DNS resource records. Please [raise an issue](https://github.com/NicTool/dns-resource-record/issues) if you cannot pass a valid resource record or you can pass an invalid resource record.

This package is for working with _individual_ Resource Records. If you're working with batches of RRs, have a look at [dns-zone](https://github.com/NicTool/dns-zone).


## USAGE

Load the index for access to all RR types:

```js
const RR = require('dns-resource-record')
```

### EXAMPLES

```js
const RR = require('dns-resource-record')
const exampleRRs = {
    A: {
        owner  : 'test.example.com',
        type   : 'A',
        address: '192.0.2.127',
        ttl    : 3600,
    },
    AAAA: {
        owner  : 'test.example.com',
        type   : 'AAAA',
        address: '2605:7900:20:a::4',
        ttl    : 3600,
    },
    SOA: {
        owner  : 'example.com',
        type   : 'SOA',
        mname  : 'matt.example.com.',
        rname  : 'ns1.example.com.',
        serial : 1,
        refresh: 7200,
        retry  : 3600,
        expire : 1209600,
        minimum: 3600,
        ttl    : 3600,
    }
}
try {
    console.log(new RR.SOA(exampleRRs.SOA))
    SOA(11) [Map] {
        'owner' => 'example.com',
        'ttl' => 3600,
        'class' => 'IN',
        'type' => 'SOA',
        'mname' => 'matt.example.com.',
        'rname' => 'ns1.example.com.',
        'serial' => 1,
        'refresh' => 7200,
        'retry' => 3600,
        'expire' => 1209600,
        'minimum' => 3600
    }
}
catch (e) {
    console.error(e.message) // invalid RRs throw
}
```

Validate records by passing a properly formatted JS object to the record-specific class. To validate an A record:

```js
const A = require('dns-resource-record').A
const validatedA = new A(exampleRRs.A)
```

Manipulate the validated record using pattern named setters:

```js
console.log(validatedA.toBind())
test.example.com    3600    IN  A   192.0.2.127

validatedA.setAddress('192.0.2.128')
console.log(validatedA.toBind())
test.example.com    3600    IN  A   192.0.2.128
```

The setters are named: `set` + `Field`, where field is the resource record field name to modify. Multi-word names are camel cased, so a field named `Certificate Usage` would have a setter named `setCertificateUsage`.

## FUNCTIONS

Get the field names for each RR type with `getFields()`:

```js
> const RR = require('dns-resource-record')
> new RR.A(null).getFields()
[ 'owner', 'ttl', 'class', 'type', 'address' ]

> new RR.PTR(null).getFields()
[ 'owner', 'ttl', 'class', 'type', 'dname' ]

> new RR.SSHFP(null).getFields()
[ 'owner', 'ttl', 'class', 'type', 'algorithm', 'fptype', 'fingerprint' ]
```

Get a list of RFCs for further learning about a RR type:

```js
> new RR.A(null).getRFCs()
[ 1035 ]

> new RR.SRV(null).getRFCs()
[ 2782 ]

> new RR.MX(null).getRFCs()
[ 1035, 2181, 7505 ]
```

### toBind

Validate a record and export to BIND format.

```js
console.log(new RR.A(exampleRRs.A).toBind())
test.example.com    3600    IN  A   192.0.2.127

console.log(new RR.AAAA(exampleRRs.AAAA).toBind())
test.example.com    3600    IN  AAAA    2605:7900:20:a::4
```

### toTinydns

Validate a record and export to tinydns format:

```js
console.log(new RR.A(exampleRRs.A).toTinydns())
+test.example.com:192.0.2.127:3600::
```

### fromTinydns toBind

Convert a tinydns line to BIND:

```js
console.log(new RR.CAA({
  tinyline: ':ns1.example.com:257:\\000\\005issue"http\\072\\057\\057letsencrypt.org":3600::\n'
}).toBind())
ns1.example.com 3600    IN  CAA 0   issue   "http://letsencrypt.org"
```

### set

The DNS validation checks can be bypassed entirely by using 'set':

```js
> validatedA.set('address', 'oops')
A(5) [Map] {
  'owner' => 'test.example.com',
  'ttl' => 3600,
  'class' => 'IN',
  'type' => 'A',
  'address' => 'oops'
}
```

Consider this a "running with scissors" mode.


## Supported Records

This module intends to include support for all current (ie, not officially deprecated) DNS RRs **and** all RRs that are in active use on the internet.

PRs are welcome, especially PRs with tests.

| **RR**     | **toBind**       | **toTinydns**    | **fromBind**     |  **fromTinydns** |
|:---------: |:----------------:|:----------------:|:----------------:|:----------------:|
| **A**      |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **AAAA**   |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **CAA**    |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **CERT**   |                  |                  |                  |                  |
| **CNAME**  |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **DNAME**  |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **DNSKEY** |:white_check_mark:|                  |:white_check_mark:|                  |
| **DS**     |:white_check_mark:|                  |:white_check_mark:|                  |
| **HINFO**  |:white_check_mark:|                  |:white_check_mark:|                  |
|**IPSECKEY**|                  |                  |                  |                  |
| **KEY**    |                  |                  |                  |                  |
| **LOC**    |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **MX**     |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **NAPTR**  |:white_check_mark:|:white_check_mark:|:white_check_mark:|                  |
| **NS**     |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **NSEC**   |                  |                  |                  |                  |
| **NSEC3**  |                  |                  |                  |                  |
| **NSEC3PARAM**|               |                  |                  |                  |
| **OPENPGPKEY**|               |                  |                  |                  |
| **PTR**    |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **RRSIG**  |                  |                  |                  |                  |
| **SIG**    |                  |                  |                  |                  |
| **SMIMEA** |:white_check_mark:|                  |:white_check_mark:|                  |
| **SOA**    |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **SPF**    |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **SRV**    |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **SSHFP**  |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **TLSA**   |:white_check_mark:|                  |:white_check_mark:|                  |
| **TXT**    |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|
| **URI**    |:white_check_mark:|:white_check_mark:|:white_check_mark:|:white_check_mark:|


## TIPS

- Domain owner names are:
    - stored fully qualified, aka absolute.
    - normalized to lower case, because:
        - DNS is case insensitive (see RFCs 4343, 1035, 1034)
        - this library enforces duplicate suppression
        - DNSSEC canonicalization (see RFC 4034)
        - wire format for most RRs require it
    - Master Zone File expansions exist at another level
- fromBIND is regex based and is naive. [dns-zone](https://github.com/NicTool/dns-zone) has a much more robust parser.
- to{Bind|MaraDNS} output can be influenced (suppress TTL, class, relative domain names) with an options object. See it in `bin/dns-zone` in the [dns-zone](https://github.com/NicTool/dns-zone) package.


## DEFINITIONS

- Resource Record: structured data associated with names / nodes
    - format: owner ttl class type rdata
    - example: www.example.com 3600 IN A 192.0.2.127
- owner [name]:
    - a node in the domain name tree
    - consists of a sequence of labels
    - format: node.zone.tld.
        - the right most label is null, aka the root
        - the 2nd from right is the top level domain
        - the 3rd from right is the [organizational] domain name
- label: character strings between the dots in a domain name
- ttl: time to live
    - how long to cache this DNS resource record
- class: IN, internet.
- type: the type of rdata the follows
    - examples: A, MX, SOA, PTR
- rdata: resource data. The contents vary widely by type
    - examples: A records have an address, CNAME records bear a cname target, NS records point to nameservers (nsdname).


## SEE ALSO

- [Wikipedia, List of DNS Record Types](https://en.wikipedia.org/wiki/List_of_DNS_record_types)


## TODO

- [x] Change all IPs to use [RFC example/doc](https://en.wikipedia.org/wiki/Reserved_IP_addresses) address space
- [x] change all domains to use reserved doc names
- [x] import tests from nictool/server/t/12_records.t
- [x] add defaults for empty values like TTL
- [x] DNSSEC RRs: DS, NSEC, NSEC3, NSEC3PARAM, RRSIG
- [x] CERT RRs: CERT, KEY, SIG, OPENPGPKEY
- [ ] APL, KX, DHCID, HIP, RP, SVCB/HTTPS
- [ ] add toWire, exports in wire/network format
- [x] RFC 4034: if the type of RR is NS, MD, MF, CNAME, SOA, MB,
      MG, MR, PTR, HINFO, MINFO, MX, RP, AFSDB, RT, SIG, PX, NXT,
      NAPTR, KX, SRV, DNAME, A6, RRSIG, or NSEC, all uppercase 
      letters in the DNS names contained within the RDATA are replaced by the lowercase letters;
- [ ] LOC record ingest/out isn't consistent with API
- [ ] handling unknown RR types: RFC 3597
- [ ] export a web page for each RR type


## DEVELOP

- this package has no dependencies. That's no accident.
- eventually this will be used a node.js app & a browser based app
    - so, ESM, eventually
- CI tests are on linux, windows, and macos


CHANGES.md

#### 1.N.N - YYYY-MM-DD


#### 0.9.8 - 2022-03-NN

-
-


#### 0.9.7 - 2022-03-29

- index
    - previousName -> previousOwner
    - export a TYPE_MAP (id => name)
- isValidHostname: allow \ char
- add the word 'RFC' in error messages citing RFCs
- when rejecting hostname, show the rejected character
- maradns: add export support
- add index.citeRFC


#### 0.9.6 - 2022-03-27

- rr\*: rename name -> owner (not overloaded)
- RFC 4034: letters in DNS names are lower cased
- README: added definitions
- repo: move from msimerson -> NicTool org
- add macos testing
- add CERT, KEY, NSEC, NSEC3, NSEC3PARAM, OPENPGPKEY, SIG


#### 0.9.5 - 2022-03-26

- README: add docs for getRFCs()
- dnskey: fix copy/paste errors
- HINFO: add fromBind
- RR.owner: if owner is same as previous, suppress (when sameOwner option is set)
- master: NODE_ENV=cov when running coverage
- TXT: pass along zone_opts


#### 0.9.4 - 2022-03-24

- add: getComment
- zone_opts, for influencing output of toBind
- normalize hostnames to lower case
- add tests: fullyQualify, getFQDN
- AAAA
    - compress: rewrote per RFC 5952, added tests
    - internally store address in expanded notation
- fromTinydns: apply correct semantics for 'x' handling
- fullyQualify
    - special handling for @
    - consider $ORIGIN
- add uc hex chars to ip6 compress


#### 0.9.3 - 2022-03-22

- hasValidLabels: remove trailing dot, else split returns empty string
- rename fullyQualified -> isFullyQualified
- rename validHostname -> isValidHostname
- set('type') no longer falls back on constructor.name (didn't reliably inherit)
- new fns: getTinyFQDN, fullyQualify, getPrefix, getFQDN
- when loading RR classes, ignore files that don't end with .js
- rr/txt: support data as array (improves idempotency)
- fromTinydns: fully qualify hostnames
- toTinydns: strip trailing . upon export
- rename getCommonFields -> getPrefixFields
- TXT: import BIND format w/o mangling WS
- SPF inherits from TXT


#### 0.9.2 - 2022-03-18

- mx: weight -> preference
- ds: keytag -> key tag (consistent naming)
- mx: add null MX test
- more use of this.getRFCs in error messages
- added getTinydnsGeneric
- on `index.is*` functions which throw, use declaratively
- class: add NONE and ANY
- validHostname: allow / char
- use \_\_dirname to find RR mods


#### 0.9.1 - 2022-03-14

- TLSA, SMIMEA: add BIND support #13


#### 0.9.0 - 2022-03-10

- added null object instantiation
- added `getTypeId`, `getDescription`, `getRdataFields`, `getQuotedFields`
- tinydns: refactored fromTinydnsGeneric
- add `RR.toBind()` for most
- spf: corrected RFC IDs
- rr/\*: refactor init into RR.constructor
- ds,dnskey: add to/from BIND support
- README: significant expansion
- tests: more signal, less noise
- tests: import tests from nictool 2


#### 0.8.1 - 2022-03-08

- use RFC example IPs and zone names


#### 0.8.0 - 2022-02-01

- fromBind for: A, AAAA, CNAME, TXT, CNAME, TXT, LOC, MX, NS, SOA, PTR, SPF, SSHFP, URI, CAA, DNAME, NAPTR
- add fromTinydns: LOC
- rr/\*: add getFields


#### 0.7.0 - 2021-10-26

- tinydns: added octalToHex, octalToUInt16, unpackDomainName, octalToInt8
- AAAA: added fromTinydns, compress, expand
- CAA, CNAME, DNAME, SPF, SSHFP, SRV, URI: added fromTinydns
- add tests for getRFCs


#### 0.6.0 - 2021-10-25

- tinydns: added octalToChar
- A, MX, NS, PTR, SOA, TXT: added fromTinydns


#### 0.5.1 - 2021-10-25

- LOC: added toTinydns
- tinydns: add UInt32toOctal
- SRV: added support


#### 0.5.0 - 2021-10-24

- NAPTR: add toTinydns
- tinydns: remove sprintf-js dependency


#### 0.4.0 - 2021-10-22

- CAA, DNAME, SSHFP, URI: add toTinydns
- lib/tinydns: added packHex and UInt16AsOctal


#### 0.3.1 - 2021-10-21

- update index.js to also export RR sub classes
- update README examples


#### 0.3.0 - 2021-10-21

- add getRFCs to all RR types
- populate this.id with IANA type ID
- toBind: use tabs for exported WS
- CAA, DNAME, NAPTR, SSHFP, URI: add toBind, tests


#### 0.2.3 - 2021-10-21

- refactored classes into separate files
- TXT, SOA, MX, CNAME, PTR, SRV: add toBind and toTinydns
- SOA: add setMinimum
- lib/tinydns: added escapeOct & packDomainName
- PTR, SRV: added tests


#### 0.2.2 - 2021-10-20

- add tests/*
- A, AAAA, add toBind and toTinydns()
- add .release


#### 0.2.1 - 2021-10-16

- additional RR formats started, weakly validated


#### 0.2.0 - 2021-10-16

- initial release & name grab


index.js

const net = require('net')

const sprintf = require('sprintf-js').sprintf

class RR extends Map {

  constructor (opts) {
    super()

    // tinydns supports these, default to empty string
    this.setLocation(opts?.location)
    this.setTimestamp(opts?.timestamp)

    this.setClass(opts?.class)
    this.setName (opts?.name)
    this.setTtl  (opts?.ttl)
    this.setType (opts?.type)
  }

  setClass (c) {
    switch (c) {
      case 'IN':
      case undefined:
        this.set('class', 'IN')
        break
      case 'CS':
      case 'CH':
      case 'HS':
        this.set('class', c)
        break
      default:
        throw new Error(`invalid class ${c}`)
    }
  }

  setLocation (l) {
    switch (l) {
      case undefined:
        this.set('location', '')
        break
      default:
        this.set('location', l)
    }
  }

  setTimestamp (l) {
    switch (l) {
      case undefined:
        this.set('timestamp', '')
        break
      default:
        this.set('timestamp', l)
    }
  }

  setName (n) {
    if (n === undefined) throw new Error(`name is required`)

    if (n.length < 1 || n.length > 255)
      throw new Error('Domain names must have 1-255 octets (characters): RFC 2181')

    this.hasValidLabels(n)

    this.set('name', n)
  }

  setTtl (t) {

    if (t === undefined) return

    if (typeof t !== 'number') throw new Error(`TTL must be numeric (${typeof t})`)

    if (parseInt(t, 10) !== t) {
      throw new Error('TTL must be a an unsigned integer')
    }

    // RFC 1035, 2181
    if (!this.is32bitInt(this.name, 'TTL', t)) return

    this.set('ttl', t)
  }

  setType (t) {
    const types = [
      'A', 'AAAA', 'CAA', 'CNAME', 'DNAME', 'LOC', 'MX', 'NAPTR', 'NS',
      'PTR', 'SSHFP', 'SOA', 'SRV', 'TXT', 'URI',
    ]
    if (!types.includes(t)) throw new Error(`type ${t} not supported (yet)`)
    this.set('type', t)
  }

  hasValidLabels (hostname) {
    for (const label of hostname.split('.')) {
      if (label.length < 1 || label.length > 63)
        throw new Error('Labels must have 1-63 octets (characters)')
    }
  }

  is8bitInt (type, field, value) {
    if (typeof value === 'number'
      && parseInt(value, 10) === value  // assure integer
      && value >= 0
      && value <= 255) return true

    throw new Error(`$type} {field} must be a 8-bit integer (in the range 0-255)`)
  }

  is16bitInt (type, field, value) {
    if (typeof value === 'number'
      && parseInt(value, 10) === value  // assure integer
      && value >= 0
      && value <= 65535) return true

    throw new Error(`$type} {field} must be a 16-bit integer (in the range 0-65535)`)
  }

  is32bitInt (type, field, value) {
    if (typeof value === 'number'
      && parseInt(value, 10) === value  // assure integer
      && value >= 0
      && value <= 2147483647) return true

    throw new Error(`$type} {field} must be a 32-bit integer (in the range 0-2147483647)`)
  }

  fullyQualified (type, blah, hostname) {
    if (hostname.slice(-1) === '.') return true

    throw new Error(`${type}: ${blah} must be fully qualified`)
  }

  validHostname (type, field, hostname) {
    if (!/[^a-zA-Z0-9\-._]/.test(hostname)) return true

    throw new Error(`${type}, ${field} has invalid hostname characters`)
  }
}

class A extends RR {
  constructor (opts) {
    super(opts)

    this.setAddress(opts.address)
  }

  setAddress (val) {
    if (!val) throw new Error('A: address is required')
    if (!net.isIPv4(val)) throw new Error('A address must be IPv4')
    this.set('address', val)
  }

  toBind () {
    return `${this.get('name')}  ${this.get('ttl')} ${this.get('class')}  ${this.get('type')} ${this.get('address')}\n`
  }

  toTinydns () {
    return `+${this.get('name')}:${this.get('address')}:${this.get('ttl')}:${this.get('timestamp')}:${this.get('location')}\n`
  }
}

class AAAA extends RR {
  constructor (opts) {
    super(opts)

    this.address(opts?.address)
  }

  address (val) {
    if (!val) throw new Error('AAAA: address is required')
    if (!net.isIPv6(val)) throw new Error('AAAA: address must be IPv6')

    this.set('address', val.toLowerCase()) // IETFs suggest only lower case
  }

  toBind () {
    return `${this.get('name')}  ${this.get('ttl')} ${this.get('class')}  ${this.get('type')} ${this.get('address')}\n`
  }

  toTinydns () {

    let val = this.get('address')

    const colons = val.match(/:/g)
    if (colons && colons.length < 7) {
      // console.log(`AAAA: restoring compressed colons`)
      val = val.replace(/::/, ':'.repeat(9 - colons.length))
    }

    // restore compressed leading zeros
    val = val.split(':').map(s => sprintf('%04s', s)).join('')

    // from AAAA notation (8 groups of 4 hex digits) to 16 escaped octals
    let rdata = ''
    for (var i = 0; i < 32; i = i+2) {
      // console.log(`i: ${i} .. ${i+1}, ${val.slice(i, i+2)}, ${parseInt(val.slice(i, i+2), 16)}`)
      rdata += sprintf('\\%03o', parseInt(val.slice(i, i+2), 16))
    }

    return `:${this.get('name')}:28:${rdata}:${this.get('ttl')}:${this.get('timestamp')}:${this.get('location')}\n`
  }
}

class CAA extends RR {
  constructor (opts) {
    super(opts)

    this.flags(opts?.flags)
    this.tags(opts?.tags)
    this.value(opts?.value)
  }

  flags (val) {
    if (!this.is8bitInt('CAA', 'flags', val)) return

    if ([ 0, 128 ].includes(val)) {
      console.warn(`CAA flags ${val} not recognized: RFC 6844`)
    }

    this.set('flags', val)
  }

  tags (val) {
    if (typeof val !== 'string'
      || val.length < 1
      || /[A-Z]/.test(val)
      || /^[a-z0-9]/.test(val))
      throw new Error('CAA flags must be a sequence of ASCII letters and numbers in lowercase: RFC 8659')

    if ([ 'issue', 'issuewild', 'iodef' ].includes(val)) {
      console.warn(`CAA tags ${val} not recognized: RFC 6844`)
    }
    this.set('tags', val)
  }

  value (val) {
    // either (2) a quoted string or
    // (1) a contiguous set of characters without interior spaces
    if (!/["']/.test(val) && /\s/.test(val)) {
      throw new Error(`CAA value may not have spaces unless quoted: RFC 8659`)
    }

    // const iodefSchemes = [ 'mailto:', 'http:', 'https:' ]
    // TODO: check if val starts with one of iodefSchemes, RFC 6844

    this.set('value', val)
  }
}

class CNAME extends RR {
  constructor (opts) {
    super(opts)

    this.address(opts?.address)
  }

  address (val) {
    if (!val) throw new Error('CNAME: address is required')

    if (net.isIPv4(val) || net.isIPv6(val))
      throw new Error(`CNAME: address must be a FQDN: RFC 2181`)

    if (!this.fullyQualified('CNAME', 'address', val)) return
    if (!this.validHostname('CNAME', 'address', val)) return
    this.set('address', val)
  }
}

class DNAME extends RR {
  constructor (opts) {
    super(opts)

    this.target(opts?.target)
  }

  target (val) {
    if (!val) throw new Error('DNAME: target is required')

    if (net.isIPv4(val) || net.isIPv6(val))
      throw new Error(`DNAME: target must be a domain name: RFC 6672`)

    if (!this.fullyQualified('DNAME', 'target', val)) return
    if (!this.validHostname('DNAME', 'target', val)) return
    this.set('target', val)
  }
}

class LOC extends RR {
  constructor (opts) {
    super(opts)

    this.address(opts?.address)
  }

  address (val) {
    // todo: validate this with https://datatracker.ietf.org/doc/html/rfc1876
    this.set('address', val)
  }
}

class MX extends RR {
  constructor (opts) {
    super(opts)

    this.address(opts?.address)
    this.weight(opts?.weight)
  }

  address (val) {
    if (!val) throw new Error('MX: address is required')

    if (net.isIPv4(val) || net.isIPv6(val))
      throw new Error(`MX: address must be a FQDN: RFC 2181`)

    if (!this.fullyQualified('MX', 'address', val)) return
    if (!this.validHostname('MX', 'address', val)) return
    this.set('address', val)
  }

  weight (val) {
    if (!this.is16bitInt('MX', 'weight', val)) return
    this.set('weight', val)
  }
}

class NAPTR extends RR {
  constructor (obj) {
    super(obj)

    this.order(obj?.order)
    this.preference(obj?.preference)
    this.flags(obj?.flags)
    this.service(obj?.service)
  }

  order (val) {
    if (!this.is16bitInt('NAPTR', 'order', val)) return

    this.set('order', val)
  }

  preference (val) {
    if (!this.is16bitInt('NAPTR', 'preference', val)) return
    this.set('preference', val)
  }

  flags (val) {
    if (![ 'S', 'A', 'U', 'P' ].includes(val))
      throw new Error (`NAPTR flags are invalid: RFC 2915`)

    this.set('flags', val)
  }

  service (val) {
    this.set('service', val)
  }
}

class NS extends RR {
  constructor (opts) {
    super(opts)

    if (!this.fullyQualified('NS', 'address', opts.address)) return
    if (!this.validHostname('NS', 'address', opts.address)) return
    this.set('address', opts.address)
  }
}

class PTR extends RR {
  constructor (obj) {
    super(obj)

    if (obj?.dname) {
      this.dname(obj?.dname)
    }
    else if (obj?.rdata) { // Generic DNS packet content
      this.dname(obj?.rdata)
    }
    else if (obj?.ptrdname) { // RFC 1035
      this.dname(obj?.ptrdname)
    }
  }

  dname (val) {
    if (!this.fullyQualified('PTR', 'dname', val)) return
    if (!this.validHostname('PTR', 'dname', val)) return

    this.set('dname', val)
  }
}

class SRV extends RR {
  constructor (opts) {
    super(opts)

    this.priority(opts?.priority)
    this.weight(opts?.weight)
    this.port(opts?.port)
    this.target(opts?.target)
  }

  priority (val) {
    if (!this.is16bitInt('SRV', 'priority', val)) return

    this.set('priority', val)
  }

  port (val) {
    if (!this.is16bitInt('SRV', 'port', val)) return

    this.set('port', val)
  }

  weight (val) {
    if (!this.is16bitInt('SRV', 'weight', val)) return

    this.set('weight', val)
  }

  target (val) {

    if (!val) throw new Error('SRV: target is required')

    if (net.isIPv4(val) || net.isIPv6(val))
      throw new Error(`SRV: target must be a FQDN: RFC 2782`)

    if (!this.fullyQualified('SRV', 'target', val)) return
    if (!this.validHostname('SRV', 'target', val)) return
    this.set('target', val)
  }
}

class SSHFP extends RR {
  constructor (opts) {
    super(opts)

    this.algorithm(opts?.algorithm)
    this.type(opts?.type)
    this.fingerprint(opts?.fingerprint)
  }

  algorithm (val) {
    // (0: reserved; 1: RSA 2: DSA 3: ECDSA 4: Ed25519 6:Ed448
    if (!this.is8bitInt('SSHFP', 'algorithm', val)) return

    this.set('algorithm', val)
  }

  type (val) {
    // 0: reserved, 1: SHA-1
    if (!this.is8bitInt('SSHFP', 'type', val)) return

    this.set('type', val)
  }

  fingerprint (val) {
    this.set('fingerprint', val)
  }
}

class SOA extends RR {
  constructor (opts) {
    super(opts)

    // name is the zone name

    // ttl, minimum (used for negative caching, since RFC 2308)
    // RFC 1912 sugggests 1-5 days
    // RIPE recommends 3600 (1 hour)

    const fields = [ 'mname', 'rname', 'serial', 'refresh', 'retry', 'expire' ]
    for (const f of fields) {
      this[f](opts[f])
    }
  }

  // MNAME (primary NS)
  mname (val) {
    if (!this.validHostname('SOA', 'MNAME', val)) return
    if (!this.fullyQualified('SOA', 'MNAME', val)) return
    this.set('mname', val)
  }

  // RNAME (email of admin)  (escape . with \)
  rname (val) {
    if (!this.validHostname('SOA', 'RNAME', val)) return
    if (!this.fullyQualified('SOA', 'RNAME', val)) return
    if (/@/.test(val)) throw new Error('SOA RNAME replaces @ with a . (dot).')
    this.set('rname', val)
  }

  serial (val) {
    if (!this.is32bitInt('SOA', 'serial', val)) return

    this.set('serial', val)
  }

  refresh (val) {
    // refresh (seconds after which to check with master for update)
    // RFC 1912 suggests 20 min to 12 hours
    // RIPE recommends 86400 (24 hours)
    if (!this.is32bitInt('SOA', 'serial', val)) return

    this.set('refresh', val)
  }

  retry (val) {
    // seconds after which to retry serial # update
    // RIPE recommends 7200 seconds (2 hours)

    if (!this.is32bitInt('SOA', 'serial', val)) return

    this.set('retry', val)
  }

  expire (val) {
    // seconds after which secondary should drop zone if no master response
    // RFC 1912 suggests 2-4 weeks
    // RIPE suggests 3600000 (1,000 hours, 6 weeks)
    if (!this.is32bitInt('SOA', 'serial', val)) return

    this.set('expire', val)
  }
}

class TXT extends RR {
  constructor (opts) {
    super(opts)

    this.address(opts?.address)
  }

  address (val) {
    this.set('address', val)
  }
}

class URI extends RR {
  constructor (opts) {
    super(opts)

    this.priority(opts?.priority)
    this.weight(opts?.weight)
    this.target(opts?.target)
  }

  priority (val) {
    if (!this.is16bitInt('URI', 'priority', val)) return

    this.set('priority', val)
  }

  weight (val) {
    if (!this.is16bitInt('URI', 'weight', val)) return

    this.set('weight', val)
  }

  target (val) {
    if (!val) throw new Error('URI: target is required')

    this.set('target', val)
  }
}

module.exports.A     = A
module.exports.AAAA  = AAAA
module.exports.CAA   = CAA
module.exports.CNAME = CNAME
module.exports.DNAME = DNAME
module.exports.LOC   = LOC
module.exports.MX    = MX
module.exports.NAPTR = NAPTR
module.exports.NS    = NS
module.exports.PTR   = PTR
module.exports.SSHFP = SSHFP
module.exports.SOA   = SOA
module.exports.SRV   = SRV
module.exports.TXT   = TXT
module.exports.URI   = URI
